### PR TITLE
rgw: Correcting logic for signature calculation for non s3 ops.

### PR DESCRIPTION
**Pull request by pritha-srivastava**
Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -3890,8 +3890,23 @@ AWSGeneralAbstractor::get_auth_data_v4(const req_state* const s,
...
+  bool is_non_s3_op = false;
+  if (s->op_type == RGW_STS_GET_SESSION_TOKEN || s->op_type == RGW_STS_ASSUME_ROLE) {
+    is_non_s3_op = true;
+  }
+
+  const char* exp_payload_hash = nullptr;
+  string payload_hash;
+  if (is_non_s3_op) {
+    //For non s3 ops, we need to calculate the payload hash
+    //const auto hash = rgw::auth::s3::calc_v4_payload_hash(s->info);
```
Comment (mattbenjamin) [inline src/rgw/rgw_rest_s3.cc]: is there a future direction for the commented case?  please elaborate

Comment (pritha-srivastava) [reply src/rgw/rgw_rest_s3.cc]: The commented part has to be removed.

File: src/rgw/rgw_rest_sts.cc [hunk 2]
```diff
@@ -201,6 +201,7 @@ RGWOp *RGWHandler_REST_STS::op_post()
            s->info.args.append(key, value);
          }
        }
+       s->info.args.append("data", post_body);
```
Comment (cbodley) [inline src/rgw/rgw_rest_sts.cc]: would it make sense to calculate the hash here and store that, instead of copying the body to hash later?

Comment (pritha-srivastava) [reply src/rgw/rgw_rest_sts.cc]: I'll check if that is possible.

Comment (pritha-srivastava) [reply src/rgw/rgw_rest_sts.cc]: @cbodley : Have made the change.

Comment (cbodley) [reply src/rgw/rgw_rest_sts.cc]: looks good!

**Comment by pritha-srivastava** [comment]
Teuthology results here: http://pulpito.ceph.com/prsrivas-2019-02-06_04:43:50-rgw-wip-rgw-awsv4signature-distro-basic-smithi/, no failures related to this PR.

**Comment by pritha-srivastava** [comment]
@mattbenjamin : This is ready to be merged.

**Comment by theanalyst** [comment]
@pritha-srivastava @mattbenjamin Do you happen to know the rgw multisite test failures in the run, I'm seeing the same in my scheduler pr teuth. runs as well. https://github.com/ceph/ceph/pull/26008

**Comment by theanalyst** [comment]
> > @pritha-srivastava @mattbenjamin Do you happen to know the rgw multisite test failures in the run, I'm seeing the same in my scheduler pr teuth. runs as well. #26008
> 
> For me I can see a OSDService destructor related stacktrace, and that is there is other test runs also, but not sure why this happens.

That might be unrelated http://tracker.ceph.com/issues/37808. The one I was referring to was particularly http://pulpito.ceph.com/prsrivas-2019-02-06_04:43:50-rgw-wip-rgw-awsv4signature-distro-basic-smithi/3555896/ which fails  tasks.rgw_multi.tests.test_versioned_object_incremental_sync 
I'm seeing the same in my branch as well

**Pull request merged**